main/src/ca/uwaterloo/flix/language/phase/Inliner.scala

```diff
@@ -209,22 +209,14 @@ object Inliner {
 
     case OccurrenceAst.Expression.Unary(sop, op, exp, tpe, purity, loc) =>
       val e = visitExp(exp, subst0)
-      (sop, e) match {
-        case (SemanticOperator.BoolOp.Not, LiftedAst.Expression.False(_)) => LiftedAst.Expression.True(loc)
-        case (SemanticOperator.BoolOp.Not, LiftedAst.Expression.True(_)) => LiftedAst.Expression.False(loc)
-        case _ => LiftedAst.Expression.Unary(sop, op, e, tpe, purity, loc)
-      }
+      val unExp = LiftedAst.Expression.Unary(sop, op, e, tpe, purity, loc)
+      unaryFold(unExp)
 
     case OccurrenceAst.Expression.Binary(sop, op, exp1, exp2, tpe, purity, loc) =>
       val e1 = visitExp(exp1, subst0)
       val e2 = visitExp(exp2, subst0)
-      (sop, e1, e2) match {
-        case (SemanticOperator.BoolOp.And, LiftedAst.Expression.False(_), _) => LiftedAst.Expression.False(loc)
-        case (SemanticOperator.BoolOp.And, _, LiftedAst.Expression.False(_)) if e1.purity == Pure => LiftedAst.Expression.False(loc)
-        case (SemanticOperator.BoolOp.Or, LiftedAst.Expression.True(_), _) => LiftedAst.Expression.True(loc)
-        case (SemanticOperator.BoolOp.Or, _, LiftedAst.Expression.True(_)) if e1.purity == Pure => LiftedAst.Expression.True(loc)
-        case _ => LiftedAst.Expression.Binary(sop, op, e1, e2, tpe, purity, loc)
-      }
+      val binExp = LiftedAst.Expression.Binary(sop, op, e1, e2, tpe, purity, loc)
```

Comment:
For both  cases, you should *not* construct a new (possibly temporary) exp just to be pass it as an argument.

In FP in general that is considered an anti pattern. One reason is that it makes it harder to understand termination. 

Instead, you should pass in each component, e,g here would be `sop`, `op`, `e1`, `e2`, and so forth, and then let binaryFold do its job.

Reply:
Ask me next time, and I will try to better explain this point :)